snuba/clickhouse/translator/visitor.py

```diff
@@ -0,0 +1,125 @@
+
+from snuba.clickhouse.query import Expression as ClickhouseExpression
+from snuba.datasets.plans.translator.visitor import (
+    ExpressionMapper,
+    MappingExpressionTranslator,
+    TranslationRules,
+)
+from snuba.query.expressions import Expression as SnubaExpression
+from snuba.query.expressions import (
+    CurriedFunctionCall,
+    ExpressionVisitor,
+    FunctionCall,
+    Lambda,
+    SubscriptableReference,
+)
+
+
+class ExpressionTranslator(MappingExpressionTranslator[ClickhouseExpression]):
+    """
+    Translates a Snuba expression into a Clickhouse expression.
+
+    As long as the Clickhouse query AST is basically the same as the Snuba one
+    we add a default rule that makes a copy of the original expression.
+    """
+
+    def __init__(
+        self, translation_rules: TranslationRules[ClickhouseExpression]
+    ) -> None:
+        default_rules = TranslationRules(
+            literals=[DefaultSimpleMapper()],
+            columns=[DefaultSimpleMapper()],
+            # TODO: remove the DefaultSubscriptableMapper translation rule as
+            # soon as we finalize the tags translation rule.
+            subscriptables=[DefaultSubscriptableMapper()],
+            functions=[DefaultFunctionMapper()],
+            curried_functions=[DefaultCurriedFunctionMapper()],
+            arguments=[DefaultSimpleMapper()],
+            lambdas=[DefaultLambdaMapper()],
+        )
+        super().__init__(
+            translation_rules=translation_rules, default_rules=default_rules
+        )
+
+
+TSimpleExp = TypeVar("TSimpleExp", bound=SnubaExpression)
+
+
+class DefaultSimpleMapper(ExpressionMapper[TSimpleExp, ClickhouseExpression]):
+    def attemptMap(
+        self,
+        expression: TSimpleExp,
+        children_translator: ExpressionVisitor[ClickhouseExpression],
+    ) -> Optional[ClickhouseExpression]:
+        return ClickhouseExpression(deepcopy(expression))
+
+
+class DefaultFunctionMapper(ExpressionMapper[FunctionCall, ClickhouseExpression]):
+    def attemptMap(
+        self,
+        expression: FunctionCall,
+        children_translator: ExpressionVisitor[ClickhouseExpression],
+    ) -> Optional[ClickhouseExpression]:
+        return ClickhouseExpression(
+            replace(
+                expression,
+                parameters=tuple(
+                    p.accept(children_translator) for p in expression.parameters
+                ),
+            )
+        )
+
+
+class DefaultCurriedFunctionMapper(
+    ExpressionMapper[CurriedFunctionCall, ClickhouseExpression]
+):
+    def attemptMap(
+        self,
+        expression: CurriedFunctionCall,
+        children_translator: ExpressionVisitor[ClickhouseExpression],
+    ) -> Optional[ClickhouseExpression]:
+        return ClickhouseExpression(
+            replace(
+                expression,
+                internal_function=expression.internal_function.accept(
+                    children_translator
+                ),
```

Comment:
Won't `accept` return a `ClickhouseExpression` which can't be assigned to `CurriedFunctionCall.internal_function` which would expect a `FunctionCall` (or `ClickhouseFunctionCall`, eventually?) Is this getting lost in `replace`?

Reply:
Actually no, Adding a new node type to achieve a strict type checking of CurriedFUnctionCall.inner_function won't work. I cannot have (in a visitor) different visit method with different return types (so that visitInnerType would return FunctionCall instead of Expression) since the return type of visitXXX is the same of the `accept` method which is defined in the abstract class Expression.

I tried a different approach which is sketched here (which is a commit in this PR that I reverted so you can still see the current version of the pr):
https://github.com/getsentry/snuba/pull/936/commits/df9097534e6b2d98c1725441419db6b2655d36c2

This is what it does:
1) TranslationRules contains also some dedicated some dedicated expression mappers to translator things like CurriedFunction.inner_function. `ExpressionMapper[FunctionCall, FunctionCall]`
2) The MappingExpressionTranslator extends the Visitor adding also a few methods to translate specific types of expressions so that they can return the type we want. Like `def translate_inner_function(self, exp: FunctionCall) -> TInnerFunctionOut`. Now TInnerFunctionOut does not need to be a generic Expression but it can be specific (FunctionCall).
3) DefaultCurriedFunctionMapper uses this dedicated method to translate internal_function. 

This guarantees strict type checking. I removed replace and it checks. The downside is the amount of type parameters I have to add. Now the ExpressionMapper takes 5 arguments. I am a bit skeptic about the complexity otherwise I would be ok cleaning that part up and bringing that commit back.
What do you think ?

Reply:
Please disregard the last comment, see this instead 
https://github.com/getsentry/snuba/pull/936#issuecomment-630702483

(leaving it here to document the previous attempt)